### ISSUE TITLE
layers and multiple series

### ISSUE THREAD

**Issue by timelyportfolio** [timelyportfolio/rcdimple#20]
To follow through on conversation with @bowerth and @bart6114 https://github.com/Bart6114/dimple/issues/1, open this issue to tackle one of the biggest missing features of `rcdimple` which is layers or multiple series.  There is already some code in place to make this happen, but the integration is poor.  I'll work up a quick prototype to start the conversation.

**Comment by timelyportfolio** [comment]
branch [feature/layers](https://github.com/timelyportfolio/rcdimple/tree/feature/layers)

**Comment by timelyportfolio** [comment]
One example from https://github.com/timelyportfolio/rcdimple/blob/feature/layers/inst/experiments/layers.R using the first draft ugly `add_series` implementation.  

```
#devtools;:install_github("timelyportfolio/rcdimple@feature/layers")
library(rcdimple)

# more testing
mtcars$name <- rownames(mtcars)
mtcars %>%
  dimple(mpg ~ cyl, type = "bar", groups = "cyl") %>%
  add_series(x = "cyl", y = "mpg", type = "line" ) %>%
  # hidden = FALSE so we can compare
  add_series(x = "cyl", y = "mpg", groups = "name", type = "bubble") %>%
  add_series(x = "cyl", y = "mpg", groups = "cyl", type = "bubble", yAxis = list(hidden=FALSE))  
```

![image](https://cloud.githubusercontent.com/assets/837910/9941472/2bc3c814-5d3b-11e5-9587-b6828607d5cb.png)

**Comment by bowerth** [comment]
Many thanks, this works quite well
haven't yet figured out how to prevent the creation of a second y-Axis on the right - is there an option to set the left-hand side y-Axis for add_series?

**Comment by timelyportfolio** [comment]
The trick (ugly I know) is `yAxis = list(hidden=FALSE)`.  I thought I had it working but it seems to be broken.  Let me dig a little.

**Comment by timelyportfolio** [comment]
The second y-axis works in these [lines](https://github.com/timelyportfolio/rcdimple/blob/feature/layers/inst/experiments/layers.R#L15-L41).

![image](https://cloud.githubusercontent.com/assets/837910/10025803/b7d7d3d6-6123-11e5-842a-0c0c7fa31d1a.png)

**Comment by timelyportfolio** [comment]
Ok  @bowerth , it appears like it the `hidden = FALSE` needs to be on the first non-primary layer.

```
mtcars$name <- rownames(mtcars)
mtcars %>%
  dimple(mpg ~ cyl, type = "bar", groups = "cyl") %>%
  add_series(x = "cyl", y = "mpg", type = "line", yAxis = list(hidden=FALSE) ) %>%
  # hidden = FALSE so we can compare
  add_series(x = "cyl", y = "mpg", groups = "name", type = "bubble") %>%
  add_series(x = "cyl", y = "mpg", groups = "cyl", type = "bubble")  
```

**Comment by bowerth** [comment]
Thanks - well, the objective is to compare on the same scale (left-hand side)
- stacked bars for breakdown in latest year
- second series (bubble) for total value in reference year

in ggplot2, this would be something like

``` r
data %>%
  filter(indic_code == "2.1.1") %>%
  ggplot(aes(x = cou)) +
  geom_bar(aes(y = value_latest, fill = var), stat = "identity") +
  geom_point(aes(y = value_earliest), stat = "identity")
```

![ggplot_example](https://cloud.githubusercontent.com/assets/2896281/10029629/50652e56-6173-11e5-9769-bfe05fb54f13.png)

**Comment by timelyportfolio** [comment]
I hate to keep showing this ugly method, but this is possible.  Unfortunately, we have to manually/programatically specify  `overrideMin` and `overrideMax` for the `yAxis`.  Here is the code for now.  I'm trying to figure out a clean way to do this, but not having much luck.

```
df %>%
  dimple( value_latest ~ cou, groups = "var", type = "bar") %>%
  xAxis( orderRule = "cou" ) %>%
  yAxis( overrideMin = 0, overrideMax = 3 ) %>%
  add_series(
    x = "cou",
    y = "value_earliest",
    type = "bubble",
    yAxis = list( overrideMin = 0, overrideMax = 3 )
  ) %>%
  add_legend()
```

![image](https://cloud.githubusercontent.com/assets/837910/10031248/c9631fa2-6141-11e5-9b37-521c93da86ba.png)

**Comment by bowerth** [comment]
many thanks for this - I'm trying to add the axis selector for addSeries to the dimple binding - see http://dimplejs.org/advanced_examples_viewer.html?id=advanced_price_range_lollipop

```
 // Add the bars mapped to the second y axis
        myChart.addSeries("Sales", dimple.plot.bar, [x, y2]);
```

for this purpose, it would be useful to check the JavaScript code that is generated by `htmlwidgets.js` and passed to `dimple-binding-0.1/dimple.js` - could you suggest a way to obtain the intermediate JS code before the SVG is generated?

**Comment by timelyportfolio** [comment]
@bowerth, same example as before?  I believe I have a solution, but I have not had a chance to implement.

**Comment by bowerth** [comment]
yes, the same example - actually, it's only one of many similar cases - the overrideMax value could be calculated from the data if necessary but adding a parameter to chose the y-axis (y1, y2) seems more intuitive. I'm looking forward to your implementation and in the meantime try to find a way to increase transparency in the JS chain `htmlwidgets.js` - `dimple.js` - `dimple.min.js`

**Comment by timelyportfolio** [comment]
I just pushed some changes in https://github.com/timelyportfolio/rcdimple/commit/e6687c0c1c157e86ac0246d9b90bce598e621bfb to allow specification of an existing axis, but I have intentionally not documented yet.  For an example, see [lines](https://github.com/timelyportfolio/rcdimple/blob/e6687c0c1c157e86ac0246d9b90bce598e621bfb/inst/experiments/layers.R#L52-L69).

In terms of transparency in the JS chain, I'm not sure I understand.  The data for an `htmlwidget` is passed in a `<script>` tag, so if you would like to see, you can do something like with Inspect Element in Chrome.

![image](https://cloud.githubusercontent.com/assets/837910/10484942/ae39a486-724d-11e5-854e-ed4e268646a0.png)

Maybe, setting a breakpoint in this [line](https://github.com/ramnathv/htmlwidgets/blob/master/inst/www/htmlwidgets.js#L444) in `htmlwidgets.js` will help.

Let me know.

**Comment by bowerth** [comment]
this works perfectly, many thanks for implementing!

**Comment by timelyportfolio** [comment]
@bowerth I only consider this a prototype for discussion.  Please let me know how it goes with additional use, so that we can refine/iterate.  Glad it worked.
